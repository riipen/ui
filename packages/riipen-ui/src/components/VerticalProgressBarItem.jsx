import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import clsx from "clsx";
import React from "react";
import PropTypes from "prop-types";
import css from "styled-jsx/css";
import _JSXStyle from "styled-jsx/style";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

import Grid from "./Grid";
import GridItem from "./GridItem";

const VerticalProgressBarItem = ({ bar, children, classes, color, icon }) => {
  const theme = React.useContext(ThemeContext);

  const getLinkedStyles = () => {
    return css.resolve`
      .icon {
        color: ${theme.palette.common.white};
        z-index: ${theme.zIndex.low};
      }
    `;
  };

  const linkedStyles = getLinkedStyles();

  return (
    <div
      className={clsx(
        "progress",
        color && bar && `progress-${color}`,
        !bar && "progress-none",
        classes
      )}
    >
      <Grid classes={[linkedStyles.className, "wrapper"]} spacing={0}>
        <GridItem lg="auto">
          <div className={clsx("iconWrapper", color && `${color}`)}>
            {!!icon && (
              <FontAwesomeIcon
                className={clsx(linkedStyles.className, "icon")}
                size="sm"
                icon={icon}
              />
            )}
          </div>
        </GridItem>
        <GridItem flexGrow={1} flexShrink={1} lg={10}>
          {children}
        </GridItem>
      </Grid>
      <style jsx>{`
        .default {
          background-color: ${theme.palette.grey[300]};
        }

        .positive {
          background-color: ${theme.palette.positive.main};
        }

        .warning {
          background-color: ${theme.palette.warning.main};
        }

        .progress {
          border-left: 8px solid var(--border-color);
          margin-left: ${theme.spacing(3)}px;
        }

        /* Progress bar colors. */
        .progress-default {
          --border-color: ${theme.palette.grey[300]};
        }

        .progress-none {
          --border-color: transparent;
        }

        .progress-positive {
          --border-color: ${theme.palette.positive.main};
        }

        .progress-warning {
          --border-color: ${theme.palette.warning.main};
        }

        .iconWrapper {
          align-items: center;
          border-radius: 50%;
          display: flex;
          height: 24px;
          justify-content: center;
          margin-bottom: ${theme.spacing(7)}px;
          margin-left: -16px;
          margin-right: ${theme.spacing(5)}px;
          margin-top: -1px;
          width: 24px;
        }
      `}</style>
      {linkedStyles.styles}
    </div>
  );
};

VerticalProgressBarItem.propTypes = {
  // external

  /**
   * The content to render inside.
   */
  children: PropTypes.node,

  /**
   * Additional classes to style with.
   */
  classes: PropTypes.array,

  /**
   * The colour of the progress icon.
   */
  color: PropTypes.oneOf(["positive", "warning", "default"]),

  /**
   * The progress icon.
   */
  icon: PropTypes.object,

  /**
   * Whether to not have bar.
   */
  bar: PropTypes.bool
};

VerticalProgressBarItem.defaultProps = {
  bar: true,
  classes: []
};

export default withClasses(VerticalProgressBarItem);
