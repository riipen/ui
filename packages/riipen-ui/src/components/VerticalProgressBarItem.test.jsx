import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import VerticalProgressBarItem from "./VerticalProgressBarItem";

describe("<VerticalProgressBarItem>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<VerticalProgressBarItem />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<VerticalProgressBarItem />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("children prop", () => {
    it("renders given children", () => {
      const child = <span>this is a child</span>;

      const wrapper = mount(
        <VerticalProgressBarItem>{child}</VerticalProgressBarItem>
      );

      expect(wrapper.text()).toContain("this is a child");
    });
  });

  describe("classes prop", () => {
    it("applies classes to the root node of VerticalProgressBarItem", () => {
      const classes = ["classOne"];

      const wrapper = mount(<VerticalProgressBarItem classes={classes} />);

      expect(
        wrapper
          .find("VerticalProgressBarItem")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClasses decorator", () => {
      const classes = ["riipen", "riipen-verticalprogressbaritem"];

      const wrapper = mount(<VerticalProgressBarItem />);

      expect(
        wrapper
          .find("VerticalProgressBarItem")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<VerticalProgressBarItem classes={classes} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("color prop", () => {
    it("renders given color", () => {
      const color = "positive";
      const wrapper = mount(<VerticalProgressBarItem color={color} />);

      expect(wrapper.find(".positive")).toHaveLength(1);
    });
  });

  describe("icon prop", () => {
    it("renders an icon", () => {
      const icon = <span />;
      const wrapper = mount(<VerticalProgressBarItem icon={icon} />);

      expect(wrapper.find("FontAwesomeIcon")).toHaveLength(1);
    });

    it("does not render an icon if not provided", () => {
      const wrapper = mount(<VerticalProgressBarItem />);

      expect(wrapper.find("FontAwesomeIcon")).toHaveLength(0);
    });
  });

  describe("bar prop", () => {
    it("passes correct class when bar and color provided", () => {
      const color = "positive";
      const wrapper = mount(<VerticalProgressBarItem color={color} />);

      expect(wrapper.find(".progress-positive")).toHaveLength(1);
    });

    it("has correct classes when bar ", () => {
      const wrapper = mount(<VerticalProgressBarItem bar={false} />);

      expect(wrapper.find(".progress-none")).toHaveLength(1);
    });
  });
});
