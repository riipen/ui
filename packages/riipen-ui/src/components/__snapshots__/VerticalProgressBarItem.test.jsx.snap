// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerticalProgressBarItem> renders correct snapshot 1`] = `
<withClasses(VerticalProgressBarItem)>
  <VerticalProgressBarItem
    bar={true}
    classes={
      Array [
        "riipen",
        "riipen-verticalprogressbaritem",
      ]
    }
  >
    <div
      className="jsx-3958355838 progress riipen riipen-verticalprogressbaritem"
    >
      <withClasses(Grid)
        classes={
          Array [
            "jsx-2412793688",
            "wrapper",
          ]
        }
        spacing={0}
      >
        <Grid
          alignItems="flex-start"
          classes={
            Array [
              "riipen",
              "riipen-grid",
              "jsx-2412793688",
              "wrapper",
            ]
          }
          flexDirection="row"
          flexGrow={1}
          flexShrink={1}
          flexWrap="wrap"
          justifyContent="flex-start"
          onResize={[Function]}
          spacing={0}
        >
          <div
            className="jsx-2773517404 riipen riipen-grid jsx-2412793688 wrapper md"
          >
            <withClasses(GridItem)
              key=".0"
              lg={2}
              size="md"
              spacing={0}
            >
              <GridItem
                alignItems="stretch"
                classes={
                  Array [
                    "riipen",
                    "riipen-griditem",
                  ]
                }
                flexDirection="column"
                flexGrow={0}
                flexShrink={0}
                flexWrap="wrap"
                justifyContent="flex-start"
                lg={2}
                size="md"
                spacing={0}
              >
                <div
                  className="jsx-3535199865 riipen riipen-griditem lg"
                >
                  <div
                    className="jsx-3958355838 iconWrapper"
                  />
                </div>
                <JSXStyle
                  dynamic={
                    Array [
                      "stretch",
                      "flex",
                      "16.666666666666664%",
                      "column",
                      0,
                      0,
                      "wrap",
                      "flex-start",
                      0,
                      0,
                    ]
                  }
                  id="1025619962"
                />
              </GridItem>
            </withClasses(GridItem)>
            <withClasses(GridItem)
              key=".1"
              lg={10}
              size="md"
              spacing={0}
            >
              <GridItem
                alignItems="stretch"
                classes={
                  Array [
                    "riipen",
                    "riipen-griditem",
                  ]
                }
                flexDirection="column"
                flexGrow={0}
                flexShrink={0}
                flexWrap="wrap"
                justifyContent="flex-start"
                lg={10}
                size="md"
                spacing={0}
              >
                <div
                  className="jsx-2265208582 riipen riipen-griditem lg"
                />
                <JSXStyle
                  dynamic={
                    Array [
                      "stretch",
                      "flex",
                      "83.33333333333334%",
                      "column",
                      0,
                      0,
                      "wrap",
                      "flex-start",
                      0,
                      0,
                    ]
                  }
                  id="1025619962"
                />
              </GridItem>
            </withClasses(GridItem)>
          </div>
          <JSXStyle
            dynamic={
              Array [
                "flex-start",
                "row",
                1,
                1,
                "wrap",
                "flex-start",
                0,
                0,
              ]
            }
            id="3466573101"
          />
        </Grid>
      </withClasses(Grid)>
      <JSXStyle
        dynamic={
          Array [
            "#dddddd",
            "#4cc88d",
            "#febb46",
            15,
            20,
            "#dddddd",
            "#4cc88d",
            "#febb46",
            10,
            35,
          ]
        }
        id="248956935"
      />
      <JSXStyle
        dynamic={
          Array [
            "#fff",
            2,
            -30,
          ]
        }
        id="2622978946"
      />
    </div>
  </VerticalProgressBarItem>
</withClasses(VerticalProgressBarItem)>
`;
